### DISCUSSION TITLE
Filter support for Haml

### DISCUSSION THREAD

**Discussion by hcthoemmes** [11ty/eleventy#3443] [Q&A]
Hi folks – this is my first time really writing code for the web (and I'm only a novice programmer in any case), and I've gotta say eleventy is making it a lot of fun ÷] There are a lot of opportunities that I wasn't aware of, one of the most interesting is definitely the use of templating languages. 

I took a gander through the ones that are presently supported, and to be honest, none of them seemed more simple to me than just writing my own HTML, save Haml. I'd like to use it for my project, but I noticed there's no support for filters. Linked questions towards the same end: are there any plans to implement that support in the near future? Is there some sort of hacky workaround to get Haml filters to work in a project? Can you use the filters of other languages? Regarding that last one: I did notice that, in the 6-minute blog tutorial, Liquid syntax was used within an HTML document, so I'm hopeful but unsure.
If I missed something in the documentation that discusses this, please point me in its direction!!

**Comment by btrem** [comment]
I don't use HAML in my project. Having a bit of experience with jinja, I went with nunjucks (which is a js port of jinja). For the most part, I use universal filters, e.g.,
```
eleventyConfig.addFilter("toArray", function(value) {
  if (Array.isArray(value)) {
    return value;
  }
  return [value];
});
```
Does that not work in Haml?

**Reply by hcthoemmes** [reply]
Hi Brian,

Haml does indeed have universal filters – but, according to the documentation, Haml's universal filters aren't supported by 11ty. I'm using Pug now, as it has a similar layout.

**Reply by btrem** [reply]
> Haml does indeed have universal filters – but, according to the documentation, Haml's universal filters aren't supported by 11ty.

When I said "universal filters," I didn't mean Haml specific filters, but rather the opposite: filters that work with any template language. It appears from reading the [docs on filters](https://www.11ty.dev/docs/filters/) that they are only available in some template languages, and haml isn't one of them.

> I'm using Pug now, as it has a similar layout.

Good to hear that you have a way to move forward.

**Comment by marnen** [comment]
This is also a deal-breaker for me in switching my site from Middleman to 11ty. I already use Haml and would like to keep it, but the easiest way for me to get certain features seems to be to use universal filters, which 11ty doesn't support for Haml.

What's the holdup in implementing this? Can I help somehow?

**Reply by sentience** [reply]
I'd imagine that the holdup is the small number of users of Haml with Eleventy, which you might attribute at least in part to limitations like this one. It's a chicken-and-egg situation.

As an outsider myself, I would suggest that the best way to help would be to look into how support for template languages are implemented in Eleventy, and see if you can contribute a PR with this missing feature yourself. I would expect that in the meantime you could use the fork, plugin, or whatever shape this implementation took to unblock your own use-case.

**Reply by Ryuno-Ki** [reply]
I mean, https://github.com/tj/haml.js wasn't updated in 12 years :-/
(Same as https://github.com/jonlunsford/middleman-haml )
Would Pug be an option?

**Reply by marnen** [reply]
@Ryuno-Ki The middleman-haml repo has nothing to do with Middleman's Haml support. The Middleman framework supports Haml templates natively (without a plugin).

Eleventy also already [officially supports Haml](https://www.11ty.dev/docs/languages/haml/), so I don't think it's unreasonable to ask for feature parity on it. Perhaps I can figure out how to implement this myself, but I thought it was worth asking before I did so.

> Would Pug be an option?

I'd use it if I absolutely had to, but I think the design decision to require a markup character for plain text and none for tag names is entirely ass-backwards (especially for a text-heavy site like mine), which is why I still prefer Haml—it works more like HTML in that respect. Plus I don't really want to convert all my existing Haml files if I can help it.

What I _might_ be able to do without modifying Eleventy, I suppose, is put the filtered stuff (which is fairly circumscribed) in a component that uses some other template language.

**Reply by marnen** [reply]
...however, I see that support for Haml was extracted into https://github.com/11ty/eleventy-plugin-template-languages, so perhaps I should take this discussion there.

**Reply by marnen** [reply]
https://github.com/11ty/eleventy-plugin-template-languages/issues/32

**Reply by marnen** [reply]
Also, Eleventy doesn't support universal filters in Pug either, so using Pug wouldn't get me what I need here in any case.

**Reply by marnen** [reply]
Pull request at https://github.com/11ty/eleventy-plugin-template-languages/pull/33
